### PR TITLE
EDU-2880: Ports 'build audiobooks from text' to golang

### PR DESCRIPTION
**Pull request by fairlydurable**
Ready for review, both text and code

### CODE REVIEW

File: docs/tutorials/go/audiobook/index.md [hunk 1]
```diff
@@ -0,0 +1,1026 @@
...
+For many people, audiobooks are a staple of daily life.
+Spoken content helps people pass time, learn new things, and stay entertained without tying themselves to a screen or a book.
+Imagine turning your daily commute, workout sessions, or household chores into opportunities to catch up on _all_ your reading -- not just books.
+You can grab text from emails, from the web, or even that report you need to "read" for work.
+With the rise of advanced text-to-speech (TTS) technology, it's not just possible, it's quick to build in Go with simple APIs.
+
+For years, text-to-speech solutions have supported the assistive community by transforming words on the page into accessible engagement.
+However, until recently, robotic voices and unnatural rhythms made TTS unappealing for casual listeners.
+If you didn't need TTS, you probably wouldn't have used TTS.
+Now, companies like [OpenAI](https://openai.com/api/) have revolutionized speech synthesis.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:26]: I assume this language already made it into the tutorial this was ported from so it's not a strong objection, but this is, imo, pitching OpenAI a little too hard

File: docs/tutorials/go/audiobook/index.md [hunk 2]
```diff
@@ -0,0 +1,1026 @@
...
+Forget about manually checking statuses or handling errors.
+Just focus on converting text while Temporal makes sure everything runs smoothly.
+When complete, you'll be able to reliably convert any plain text file into an audiobook.
+
+Ready to transform your text into immersive audiobooks?
+Get started by checking that you have the necessary understanding, tools, and environment set up.
+
+## Prerequisites
+
+You can build the project by following this tutorial or, if you're feeling impatient, just grab the ready-to-go source from its [GitHub repo](https://github.com/temporalio/build-audiobook-go).
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:42]:
```suggestion
You can build the project by following this tutorial, or just grab the ready-to-go source from its [GitHub repo](https://github.com/temporalio/build-audiobook-go).
```

I like the editorializing but I would tone it down here

File: docs/tutorials/go/audiobook/index.md [hunk 3]
```diff
@@ -0,0 +1,1026 @@
...
+If you want to play first and explore later, you can come back and read through the how-to and background.
+Here's what you need to get going:
+
+1. **A local Go development environment**:
+   Follow ["Set up a local development environment"](/getting_started/go/dev_environment/) so you're ready to build Temporal applications with Go.
+   Ensure a local Temporal Service is running and that you can access the Temporal Web UI from port `8080`.
+   These services are necessary for you to build and run this project.
+
+1. **Basic understanding of the Temporal Go SDK**:
+   Work through the [Hello World in Go](/getting_started/Go/hello_world_in_Go/) tutorial.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:53]:
```suggestion
   Work through the [Hello World in Go](/getting_started/go/hello_world_in_go/) tutorial.
```

File: docs/tutorials/go/audiobook/index.md [hunk 4]
```diff
@@ -0,0 +1,1026 @@
...
+   This covers the basics of getting a Temporal Go SDK project up and running.
+
+1. **OpenAI API access**:
+   Sign up for your [OpenAI API developer account](https://openai.com/api/) and create a bearer token.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:57]: Can we link directly to their docs for creating a token? It would be scope creep to walk through all of it here, but just linking to their homepage doesn't feel like enough.

File: docs/tutorials/go/audiobook/index.md [hunk 5]
```diff
@@ -0,0 +1,1026 @@
...
+   You need this token to access OpenAI services for text-to-speech conversion in your project.
+
+Once caught up with prerequisites, you can build out your Go project directory.
+
+:::info CAUTIONS
+
+This is a tutorial project and its implementation is suited for personal and hobbyist use.
+In production, you wouldn't read or write from a single database file, hard drive, or system.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:65]: I would rewrite this a bit -- the mention of an individual "hard drive" is unnecessary, and SQLite is actually used in production lots of places :)

File: docs/tutorials/go/audiobook/index.md [hunk 6]
```diff
@@ -0,0 +1,1026 @@
...
+   .PHONY: all clean build build-worker clean-deps fetch-deps worker
+   ```
+
+4. Create `go.mod` by entering:
+
+   ```
+   go mod init audiobook/app
+   ```
+   
+   Edit your new `go.mod` file to add two requirements:
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:135]: Missing a step number here, and it would maybe be helpful to show what `go.mod` looks like after being created but before being appended to, it's a bit confusing for first-timers.

File: docs/tutorials/go/audiobook/index.md [hunk 7]
```diff
@@ -0,0 +1,1026 @@
...
+       github.com/stretchr/testify v1.8.0
+   )
+   ```
+   
+   Your dependencies include:
+   
+   * **go.temporal.io/sdk v1.17.0**: Temporal's open source failure mitigation Go SDK.
+   * **github.com/stretchr/testify v1.8.0**: Write tests with assertions and mocks.
+
+5. Establish your bearer token environment variable
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:153]:
```suggestion
5. Establish your bearer token environment variable.
```

File: docs/tutorials/go/audiobook/index.md [hunk 8]
```diff
@@ -0,0 +1,1026 @@
...
+   Your application uses the environment variable to authenticate with the OpenAI service.
+   It needs to live in the same shell as  your running app.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:155]: "live in the same shell" is potentially very unclear to junior devs who don't think much about env variables, and I definitely don't think we need to document tcsh -- almost every beginner tutorial like this I've seen defaults to bash syntax. I would simplify all of this and just focus on prepending the `export` syntax later on when it's actually used.

File: docs/tutorials/go/audiobook/index.md [hunk 9]
```diff
@@ -0,0 +1,1026 @@
...
+   setenv OPEN_AI_BEARER_TOKEN 'your-secret-bearer-token'
+   ```
+
+   or bash:
+
+   ```
+   export OPEN_AI_BEARER_TOKEN='your-secret-bearer-token'
+   ```
+
+   Optionally, create a `bearer.sh` utility to reduce typing.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:169]: I would cut this, most people aren't going to find it simpler to read about creating shell scripts, and those who already know how to do so, can.

File: docs/tutorials/go/audiobook/index.md [hunk 10]
```diff
@@ -0,0 +1,1026 @@
...
+
+if chunk.Len() > 0 {
+    chunks = append(chunks, chunk.String())
+}
+
+return chunks, nil
+
+```
+
+As a rule-of-thumb, a typical English word uses 1.33 OpenAI tokens, which is how this code defines the `AVERAGE_TOKENS_PER_WORD` constant.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:486]: I would swap this paragraph and the following one for clarity

File: docs/tutorials/go/audiobook/index.md [hunk 11]
```diff
@@ -0,0 +1,1026 @@
...
+
+Tokens quantify the data processed by OpenAI requests and all OpenAI endpoints use token limits.
+This code creates chunks with approximately 512 token for each API request.
+Although the OpenAI token limit is higher than this, this code is conservative to reduce bandwidth for the returned audio bytes.
+
+After building a method to process string data, you'll add code to build an output file in the next steps.
+
+### Build a temporary file to store output with `CreateTemporaryFile`
+
+The `CreateTemporaryFile` activity requests that the system create a new temporary file.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:497]: By now, I'm confused whether I should be appending these additional functions to an in-progress script, or whether you just wanted me to copy and paste the whole darn thing above and are just walking through it now. Since you've gone to the trouble of walking through it step by step already, I would remove the whole file contents up above and rewrite that part so it's clear you'll be building it incrementally throughout the tutorial.

File: docs/tutorials/go/audiobook/index.md [hunk 12]
```diff
@@ -0,0 +1,1026 @@
...
+   └── worker
+   ```
+   
+2. Move to the root directory:
+   
+  ```sh
+   cd TTSWorker
+   ```
+
+3. Create your Makefile in the `TTSWorker/` project root directory.
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:94]: I am not too comfortable with the use of Makefiles for Go projects -- in all the rest of our docs, courses, and tutorials, we just use `go run`. I could be talked around on this, but I think this adds complexity that is inconsistent with our house style and doesn't make learners' lives much easier.

File: docs/tutorials/go/audiobook/index.md [hunk 13]
```diff
@@ -0,0 +1,1026 @@
...
+
+### Run the Worker application
+
+1. Update your dependencies:
+
+   ```sh
+   go mod tidy
+   ```
+
+1. Build the Worker app:
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:895]: See earlier point about Makefiles

File: docs/tutorials/go/audiobook/index.md [hunk 14]
```diff
@@ -0,0 +1,1026 @@
...
+The results path is also displayed as part of the CLI's `workflow execute` command output and in the Worker logs.
+
+:::tip Cautions and notes
+
+- Do not modify your input or output files while the Workflow is running.
+- The Workflow fails if you don't pass a valid text file named with a `txt` extension.
+
+:::
+
+### Check your progress
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:961]: This part, and the above CLI submission step, are especially great -- would love to see them made more prominent, I think simplifying other parts of this tutorial will achieve that.

**Comment by axfelix** [comment]
Looks great! Haven't done an end-to-end test yet, but left some comments, mostly on scope.

File: docs/tutorials/go/audiobook/index.md (hunk 1 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:26]: Brian said "let OpenAI be the star" (more or less).

File: docs/tutorials/go/audiobook/index.md (hunk 2 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:42]: Adopting.

File: docs/tutorials/go/audiobook/index.md (hunk 3 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:53]: yeeps. Fixing.

File: docs/tutorials/go/audiobook/index.md (hunk 4 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:57]: They're re-organizing from individual to organizational so I don't want to link to things that are changing. But very nice call. Thank you.

File: docs/tutorials/go/audiobook/index.md (hunk 5 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:65]: This is actually by Chad, with a tiny bit of commentary by Mason and Brian, and some minor edits from me. Removing hard drive.

File: docs/tutorials/go/audiobook/index.md (hunk 6 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:135]: Adopted.

File: docs/tutorials/go/audiobook/index.md (hunk 7 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:153]: Adopted

File: docs/tutorials/go/audiobook/index.md (hunk 8 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:155]: I'll change it to bash, and I'm going to use terminal session.

File: docs/tutorials/go/audiobook/index.md (hunk 9 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:169]: Removing.

File: docs/tutorials/go/audiobook/index.md (hunk 10 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:486]: swapped

File: docs/tutorials/go/audiobook/index.md (hunk 11 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:497]: Brian said: build the file first, explain after. I'll make that clearer for each section.

File: docs/tutorials/go/audiobook/index.md [hunk 15]
```diff
@@ -0,0 +1,1026 @@
...
+
+```sh
+temporal server start-dev \
+    --ui-port 8080 \
+    --db-filename /path/to/your/temporal.db
+```
+
+Once running, connect to the [Temporal Web UI](http://localhost:8080/) and verify that the server is working.
+
+### Instantiate Your Bearer Token
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:881]: Removing from above.

File: docs/tutorials/go/audiobook/index.md (hunk 12 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:94]: Make removed

File: docs/tutorials/go/audiobook/index.md (hunk 13 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:895]: ditto

File: docs/tutorials/go/audiobook/index.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (axfelix) [reply docs/tutorials/go/audiobook/index.md:819]: Oh, sorry, I think you cut more than I intended -- I'd like to see an example that actually looks like `OPEN_AI_BEARER_TOKEN=xyz go run bin/worker` in the "Instantiate Your Bearer Token" section so people know exactly what to run.

File: docs/tutorials/go/audiobook/index.md [hunk 16]
```diff
@@ -0,0 +1,971 @@
...
+- Finally, it runs the Worker, which starts polling for Tasks.
+
+Your complete project structure will now look like this:
+
+```sh
+.
+├── TTSActivities.go
+├── TTSWorkflow.go
+├── go.mod
+├── go.sum
```
Comment (axfelix) [inline docs/tutorials/go/audiobook/index.md:794]: Very minor point, but `go.sum` shouldn't actually exist at this stage, since you haven't run `go mod tidy` to download dependencies yet

File: docs/tutorials/go/audiobook/index.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (axfelix) [reply docs/tutorials/go/audiobook/index.md:819]: Also -- you shouldn't need the intermediate `go build` step if you replace the third one with `go run`.

File: docs/tutorials/go/audiobook/index.md (hunk 16 above)
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:794]: Fixing!

File: docs/tutorials/go/audiobook/index.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/go/audiobook/index.md:819]: Resolved. Discussed. Adopted.

**Review by axfelix — APPROVED** [review]
